python/cudf/cudf/_libxx/join.pyx

```diff
@@ -0,0 +1,155 @@
+from cudf._libxx.lib cimport *
+from cudf._libxx.table cimport Table, columns_from_ptr
+from cudf._libxx.includes.join cimport *
+from libcpp.vector cimport vector
+from libcpp.pair cimport pair
+
+from collections import OrderedDict
+
+cpdef join(Table lhs,
+           Table rhs,
+           object left_on,
+           object right_on,
+           object how,
+           object method,
+           bool left_index=False,
+           bool right_index=False
+           ):
+    """
+    Call libcudf++ join for full outer, inner and left joins.
+    Returns a list of tuples [(column, valid, name), ...]
+    """
+
+    cdef Table c_lhs = lhs
+    cdef Table c_rhs = rhs
+
+    cdef vector[int] left_on_ind
+    cdef vector[int] right_on_ind
+    num_inds_left = len(left_on) + left_index
+    num_inds_right = len(right_on) + right_index
+    left_on_ind.reserve(num_inds_left)
+    right_on_ind.reserve(num_inds_right)
+
+    columns_in_common = OrderedDict()
+    cdef vector[pair[int, int]] c_columns_in_common
+
+    # Views might or might not include index
+    cdef table_view lhs_view
+    cdef table_view rhs_view
+
+    # the result columns are all the left columns (including common ones)
+    # + all the right columns (excluding the common ones)
+    result_col_names = [None] * len(lhs._data.keys() | rhs._data.keys())
+
+    ix = 0
+    for name in lhs._data.keys():
+        result_col_names[ix] = name
+        ix += 1
+    for name in rhs._data.keys():
+        if name not in lhs._data.keys():
+            result_col_names[ix] = name
+            ix += 1
+
+    # keep track of where the desired index column will end up
+    result_index_pos = None
+
+    if left_index or right_index:
+        # If either true, we need to process both indices as columns
+        lhs_view = c_lhs.view()
+        rhs_view = c_rhs.view()
+
+        left_join_cols = [lhs.index.name] + list(lhs._data.keys())
+        right_join_cols = [rhs.index.name] + list(rhs._data.keys())
+        if left_index and right_index:
+            left_on_idx = right_on_idx = 0
+            result_index_pos = 0
+            result_index_name = rhs.index.name
+
+        elif left_index:
+            left_on_idx = 0
+            right_on_idx = right_join_cols.index(right_on[0])
+            result_index_pos = len(left_join_cols)
+            result_index_name = rhs.index.name
+
+            # common col gathered from the left
+            common = result_col_names.pop(result_col_names.index(right_on[0]))
+            result_col_names = [common] + result_col_names
+        elif right_index:
+            right_on_idx = 0
+            left_on_idx = left_join_cols.index(left_on[0])
+            result_index_pos = 0
+            result_index_name = lhs.index.name
+
+        left_on_ind.push_back(left_on_idx)
+        right_on_ind.push_back(right_on_idx)
+
+        columns_in_common[pair[int, int](
+            (left_on_idx, right_on_idx)
+        )] = None
```

Comment:
@brandon-b-miller Is the key of the OrderedDict a `pair[int, int]`? Does this get converted into a Python object implicitly? Seems like we should just use a tuple here instead of relying on Cython to do implicit conversion for us unnecessarily.